### PR TITLE
[2374] - feat(course preview) - display financial incentives

### PR DESCRIPTION
**Pull request by TobyRet**
### Context

The course preview page (`/organisations/[PROVIDER_CODE]/2020/courses/[COURSE_CODE]/preview`)will now display financial incentives (scholarship, bursary or early career payments) depending on whether this information is returned from the backend API

This replicates what is shown on `search-and-compare-ui`

### Changes proposed in this pull request
- Creation of financial incentives partials
- Introduces new CourseDecorator logic to determine which financial incentives are displayed

### Screenshots

Below are screen grabs of course previews and the range of financial incentives.

**1. 'Salaried' course**

salary

**2. 'Primary with Mathematics' course**

primary_with_mathematics

**3. Course with bursary**

bursary

**4. Course with bursary and scholarship ( and / or early career payments)**

scholarship_and_bursary

**5. Course with finance only**

loan

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by TobyRet** [comment]
[Early Career Payment Link is hardcoded](https://trello.com/c/ronCdOgx/2553-early-career-payment-link-is-hardcoded)

**Comment by defong** [comment]
did you change the bit at the top as well

**Comment by TobyRet** [comment]
> image
> 
> refereneced
> 
> https://find-postgraduate-teacher-training.education.gov.uk/results/filter/subject?subjects=3&qualifications=QtsOnly,PgdePgceWithQts,Other&fulltime=False&parttime=False&hasvacancies=True&senCourses=False
> 
> why £3500

This was old copy in the partial I re-enabled. 
Will update.

**Comment by TobyRet** [comment]
> image
> 
> did you change the bit at the top as well

Nope. I'll update

**Comment by TobyRet** [comment]
@defong - funding options fixed

Screenshot 2019-11-19 at 09 29 11

File: app/decorators/course_decorator.rb [hunk 1]
```diff
@@ -46,6 +46,49 @@ def funding
...
+      object.name
+    end
+  end
+
+  def has_scholarship_and_bursary?
+    has_bursary? && has_scholarship?
+  end
+
+  def has_scholarship_or_bursary?
+    has_bursary? && has_scholarship?
```
Comment (davidgisbey) [inline app/decorators/course_decorator.rb]: `has_bursary? || has_scholarship?`

**Comment by davidgisbey** [comment]
Looks good to me once the boolean is corrected and Sau's comments are addressed 👍

File: app/decorators/course_decorator.rb (hunk 1 above)
Comment (TobyRet) [reply app/decorators/course_decorator.rb]: Woops!

File: spec/decorators/course_decorator_spec.rb [hunk 2]
```diff
@@ -197,4 +197,131 @@
...
+    describe "#funding_option" do
+      let(:subject) { decorated_course.funding_option }
+
+      context "Salary" do
+        let(:course) { build :course, funding_type: "salary" }
+
+        it { is_expected.to eq("Salary") }
+      end
+
+      context "Bursasry and Scholarship" do
```
Comment (misaka) [inline spec/decorators/course_decorator_spec.rb]:
```suggestion
      context "Bursary and Scholarship" do
```

File: app/decorators/course_decorator.rb [hunk 3]
```diff
@@ -162,4 +201,14 @@ def not_on_find
   def edit_vacancy_link
     h.link_to("Edit", h.vacancies_provider_recruitment_cycle_course_path(provider_code: object.provider_code, recruitment_cycle_year: object.recruitment_cycle_year, code: object.course_code), class: "govuk-link")
   end
+
+  def find_max(attribute)
+    subject_attributes = object.subjects.map do |s|
+      if s.attributes[attribute].present?
+        s.send(attribute).to_i
```
Comment (defong) [inline app/decorators/course_decorator.rb]: `send` vs `__send__` ?

File: app/views/courses/preview/financial_support/_scholarship_and_bursary.html.erb [hunk 4]
```diff
@@ -1,26 +1,34 @@
...
+
+      <p class="govuk-body">
+        You can’t claim both a bursary and a scholarship - you can only claim one.
+      </p>
+
+      <p class="govuk-body">
+        Find out how to <%= link_to 'apply for a scholarship', 'https://getintoteaching.education.gov.uk/funding-and-salary/overview/scholarships', class: 'govuk-link' %>. You don’t need to apply for a bursary - if you’re eligible, you’ll automatically start receiving it once you begin your course. Find out <%= link_to 'how you’ll be paid', 'https://getintoteaching.education.gov.uk/funding-and-salary/overview/how-you-will-be-paid', class: 'govuk-link' %>.
+      </p>
+    </div>
+  <% end %>
```
Comment (defong) [inline app/views/courses/preview/financial_support/_scholarship_and_bursary.html.erb]: https://github.com/DFE-Digital/search-and-compare-ui/blob/e26b23f3b803be59e82c1f0763402d01b1713bd9/shared/Views/Shared/Components/CourseDetails/FinancialSupport/_FinancialSupport.cshtml#L29-L32

image

Other Loans??

https://github.com/DFE-Digital/search-and-compare-ui/blob/e26b23f3b803be59e82c1f0763402d01b1713bd9/shared/Views/Shared/Components/CourseDetails/FinancialSupport/_FinancialSupport.cshtml#L35

image

what happens if you outside uk

File: app/views/courses/preview/financial_support/_scholarship_and_bursary.html.erb [hunk 5]
```diff
@@ -1,26 +1,34 @@
...
-<ul class="govuk-list govuk-list--bullet">
-  <li data-qa="course__scholarship_amount">a scholarship of <%= number_to_currency(course.scholarship_amount) %></li>
-  <li data-qa="course__bursary_amount">a bursary of <%= number_to_currency(course.bursary_amount) %></li>
-</ul>
-
-<% if course.has_early_career_payments? %>
+<% if course.has_financial_incentives? %>
   <p class="govuk-body">
-    With a scholarship or bursary, you’ll also get early career payments of £5,000 each in your third and fifth year of teaching (£7,500 in <%= link_to 'some areas of England', 'https://www.gov.uk/guidance/mathematics-early-career-payments-guidance-for-teachers-and-schools', class: "govuk-link" %>).
+    You could be eligible for either:
```
Comment (defong) [inline app/views/courses/preview/financial_support/_scholarship_and_bursary.html.erb]: https://find-postgraduate-teacher-training.education.gov.uk/course/2AT/2T89#section-financial-support

https://github.com/DFE-Digital/search-and-compare-ui/blob/master/shared/Views/Shared/Components/CourseDetails/FinancialSupport/_Bursary.cshtml

image

^^ should we still be presenting like the above?

File: app/views/courses/preview/financial_support/_financial_support.html.erb [hunk 6]
```diff
@@ -4,13 +4,11 @@
   <% if course.funding_type == 'salary' %>
     <%= render partial: 'courses/preview/financial_support/salaried' %>
   <% else %>
-    <p class="app-missing-section">
-      You can’t preview information about bursaries, scholarships and financial support, but it will be displayed in Find postgraduate teacher training.
-    </p>
+    <%= render partial: 'courses/preview/financial_support/scholarship_and_bursary' %>
```
Comment (defong) [inline app/views/courses/preview/financial_support/_financial_support.html.erb]: You probably better off porting across this 

https://github.com/DFE-Digital/search-and-compare-ui/blob/master/shared/Views/Shared/Components/CourseDetails/FinancialSupport/_FinancialSupport.cshtml

**Comment by defong** [comment]
Don't know how much like for like is needed or where to draw the line. 

What's implemented is mainly `scholarship` and `bursary` where its fine to show a bullet list with 2 items, that's mainly good to go, minus default text not shown

It will look odd to have a a bullet list for one item.

From a data point of view it's probably a given that if there is a `scholarship` then it will have a `bursary`.

File: app/decorators/course_decorator.rb (hunk 3 above)
Comment (TobyRet) [reply app/decorators/course_decorator.rb]: https://stackoverflow.com/questions/4658269/ruby-send-vs-send

Don't think there is any issue with using 'send' in this instance but will change for consistency

Comment (defong) [reply app/decorators/course_decorator.rb]: No idea just highlighting

File: app/decorators/course_decorator.rb [hunk 7]
```diff
@@ -100,10 +139,10 @@ def has_site?(site)
   def funding_option
     if object.funding_type == "salary"
       "Salary"
-    elsif object.has_scholarship_and_bursary?
+    elsif has_scholarship_and_bursary?
       "Scholarship, bursary or student finance if you’re eligible"
```
Comment (misaka) [inline app/decorators/course_decorator.rb]: I understand that this is existing code, but is it correct? It seems quite possible that courses will only have one or the other, so:

Q1. should only courses that have both behave this text?
Q2. is "student finance" the same as "early_career_payments"? Just trying to match the wording here to what's in the DB, where we have `bursary_amount` and `scholarship` and `early_career_payments`, but no "student finance".
Q3. should we move this to locales?

Comment (defong) [reply app/decorators/course_decorator.rb]: Q2: no they not the same thing

File: app/decorators/course_decorator.rb [hunk 8]
```diff
@@ -46,6 +46,68 @@ def funding
...
+    if requirements.count > 1
+      ":"
+    else
+      "#{requirements.first}."
+    end
+  end
+
+  def requirements
+    requirements = ["a degree of 2:2 or above in any subject"]
+    mathematics_requirement = "at least grade B in maths A-level (or an equivalent)"
```
Comment (defong) [inline app/decorators/course_decorator.rb]: move to `if` block?

File: spec/decorators/course_decorator_spec.rb [hunk 9]
```diff
@@ -197,4 +197,202 @@
...
+      context "course has other financial incentives apart from bursaries" do
+        let(:mathematics) { build :subject, bursary_amount: "2000" }
+        let(:english) { build :subject, scholarship: "4000" }
+        let(:subjects) { [mathematics, english] }
+
+        it { is_expected.to_not be_bursary_only }
+      end
+    end
+
+    describe "Bursaries" do
```
Comment (defong) [inline spec/decorators/course_decorator_spec.rb]: not sure about this?

maybe is a context about all `bursary` like you did with `scholarships`

File: app/views/courses/preview/_financial_incentives.html.erb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+<% if course.salaried? %>
+  <%= render partial: 'courses/preview/financial_support/salaried' %>
+<% elsif course.bursary_only? %>
+  <%= render partial: 'courses/preview/financial_support/bursary' %>
+<% elsif course.has_scholarship_and_bursary? %>
+  <%= render partial: 'courses/preview/financial_support/scholarship_and_bursary' %>
+<% else %>
+  <%= render partial: 'courses/preview/financial_support/loan' %>
```
Comment (defong) [inline app/views/courses/preview/_financial_incentives.html.erb]: Boilerplate for all but salaried

```ruby 
if salaried?
  ...
else
  if
    ...
  else
    ...
  end
#boilerplating
	
end
```

File: app/decorators/course_decorator.rb [hunk 11]
```diff
@@ -46,6 +46,68 @@ def funding
...
+
+  def bursary_first_line_ending
+    if requirements.count > 1
+      ":"
+    else
+      "#{requirements.first}."
+    end
+  end
+
+  def requirements
```
Comment (defong) [inline app/decorators/course_decorator.rb]: can we call this something else as
https://find-postgraduate-teacher-training.education.gov.uk/course/T92/X130#section-entry

on the course details page there is already something called `requirements`

maybe financial_incestive_requirements

**Comment by defong** [comment]
> ### Guidance to review
> 
>     * In the 'early career payments` section the subject name displayed depends on how may subjects a course has. If more than one then it reverts to the course name (discussed with @defong)

Misquoted, but since you mentioned 'early career payments`, I doubled checked

- if the course has two subjects and they are `Mathematics` & `Physics` secondary subjects then there is no 'early career payments`

https://github.com/DFE-Digital/search-and-compare-ui/blob/34ca77118ebd756ec056155f4c7093646355f930/shared/ViewModels/FinanceViewModel.cs#L70-L83

But ignore that for now as I don't think its right @dankmitchell any idea?
image


This was the bit I was refering to 
HasBursary & HasScholarshipAndBursary is overridden if 

course name starts with and has 2 subjects associated with it 
```
                "Drama",
                "Media Studies",
                "PE",
                "Physical Education",
```

see
https://github.com/DFE-Digital/search-and-compare-ui/blob/903e35ea800ccc4cf0b06a18a36e037ab813e229/shared/ViewFormatters/CourseExtensions.cs#L109-L158

File: app/decorators/course_decorator.rb (hunk 8 above)
Comment (TobyRet) [reply app/decorators/course_decorator.rb]: `if` block is on line 73 - that what you mean?

File: app/views/courses/preview/_financial_incentives.html.erb (hunk 10 above)
Comment (TobyRet) [reply app/views/courses/preview/_financial_incentives.html.erb]: Salaried partial is on line 2

```
<% if course.salaried? %>
  <%= render partial: 'courses/preview/financial_support/salaried' %>
<% elsif course.bursary_only? %>
  <%= render partial: 'courses/preview/financial_support/bursary' %>
<% elsif course.has_scholarship_and_bursary? %>
  <%= render partial: 'courses/preview/financial_support/scholarship_and_bursary' %>
<% else %>
  <%= render partial: 'courses/preview/financial_support/loan' %>
<% end %>
```

I'm assuming the above is ok as a non-salaried course would not have a bursary or scholarship right?

File: app/decorators/course_decorator.rb (hunk 11 above)
Comment (TobyRet) [reply app/decorators/course_decorator.rb]: Uh I'm not so sure about `financial_incestive_requirements`...
... `financial_incentive_requirements` might be more family-friendly

File: app/decorators/course_decorator.rb (hunk 8 above)
Comment (defong) [reply app/decorators/course_decorator.rb]: mathematics_requirement only used in line 74 right? move closer to usage?

File: spec/views/preview.html.erb_spec.rb [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+  context "course has no financial incentives" do
+    it "renders the 'loan' partial" do
+      course = build(:course)
+
+      render "courses/preview/financial_support", course: course.decorate
+
+      preview_course_page.load(rendered)
+
+      expect(preview_course_page).to have_selector("[data-qa=course__loan_details]")
```
Comment (TobyRet) [inline spec/views/preview.html.erb_spec.rb]: I wonder if I need to asset _other_ partials are _not_ appearing as well? Possibly overkill ...

Comment (TobyRet) [reply spec/views/preview.html.erb_spec.rb]: If these view specs provide little value then let me know

File: spec/site_prism/page_objects/page/organisations/course_preview.rb [hunk 13]
```diff
@@ -24,6 +24,8 @@ class CoursePreview < PageObjects::Base
         element :salary_details, "#section-salary"
         element :scholarship_amount, "[data-qa=course__scholarship_amount]"
         element :bursary_amount, "[data-qa=course__bursary_amount]"
+        element :early_career_payment_details, "[data-qa=course__early_career_payment_details]"
```
Comment (defong) [inline spec/site_prism/page_objects/page/organisations/course_preview.rb]: Can we add the others here?

Comment (TobyRet) [reply spec/site_prism/page_objects/page/organisations/course_preview.rb]: This is a PageObject used for testing, therefore the only elements that should appear hear are the ones that used in the associated spec. The feature/preview_spec.rb tests the happy path only with a course that has scholarship, bursary and early career payment incentives.

Comment (defong) [reply spec/site_prism/page_objects/page/organisations/course_preview.rb]: so `"[data-qa=course__loan_details]"` from
spec/views/preview.html.erb_spec.rb

**Pull request merged**